src/passes/OptimizeInstructions.cpp

```diff
@@ -1204,6 +1207,29 @@ struct OptimizeInstructions
       }
     }
 
+    auto features = getModule()->features;
+
+    // Repeated identical ref.cast operations are unnecessary, if using the
+    // exact same rtt - the result will be the same. Find the immediate child
+    // cast, if there is one, and see if it is identical.
+    // TODO: Look even further through incompatible casts?
+    auto* ref = curr->ref;
+    while (!ref->is<RefCast>()) {
+      auto* last = ref;
+      ref = Properties::getImmediateFallthrough(ref, passOptions, features);
+      if (ref == last) {
+        break;
+      }
+    }
+    if (auto* child = ref->dynCast<RefCast>()) {
+      // Check if the casts are identical.
+      if (ExpressionAnalyzer::equal(curr->rtt, child->rtt) &&
```

Comment:
Would it be worth checking instead whether the current rtt is a sub-rtt of the child's rtt? Then we could skip the child cast and just keep the current cast. That optimization would be more general than this optimization where they have to be identical.

Reply:
To find out if an RTT is a sub-RTT of another we need more than the static types, though. Even if two rtts have identical types, the cast may or may not fail, e.g. these two have identical type `(rtt 2 $C)` but they cannot be cast:
```wat
(rtt.sub $C (rtt.sub $B (rtt.canon $A)))
(rtt.sub $C (rtt.sub $B (rtt.canon $B)))
```

Unless I'm missing something, handling the general case doesn't seem practical. (And is one reason why RTTs make optimization hard at the wasm level IMO.)

Reply:
Ugh right, that still always confuses me. This seems like a good reason to get rid of first-class RTTs for nominal types.